src/middlewared/middlewared/utils/io.py

```diff
@@ -1,24 +1,124 @@
+
+    `uid` - expected numeric UID for file
+
+    `gid` - expected numeric GID for file
+
+    `perms` - numeric permissions that file should have
+
+    `raise_error` - raise an exception if file ownership or permissions have unexpectedly
+    changed.
+    """
 
     if isinstance(data, str):
         data = data.encode()
 
-    if isinstance(path, int):
-        path = f'/proc/self/fd/{path}'
+    if not isinstance(perms, int):
+        raise ValueError('perms must be an integer')
+
+    if perms & ~(stat.S_IRWXU | stat.S_IRWXG | stat.S_IRWXO):
+        raise ValueError(f'{perms}: invalid mode')
+
+    for xid in ((uid, 'uid'), (gid, 'gid')):
+        value, name = xid
+        if not isinstance(value, int):
+            raise ValueError(f'{name} must be an integer')
+
+        if value < 0 or value > ID_MAX:
+            raise ValueError(f'{name} must be between 0 and {ID_MAX}')
+
+    if dirfd is not None:
+        if not isinstance(dirfd, int):
+            raise ValueError('dirfd must be a valid file descriptor')
+
+        if not os.path.exists(f'/proc/self/fd/{dirfd}'):
+            raise ValueError(f'{dirfd}: file descriptor not found')
+
+    if dirfd is not None:
+        # tempfile API does not permit using a file descriptor
+        parent_dir = os.readlink(f'/proc/self/fd/{dirfd}')
+    else:
+        parent_dir = os.path.dirname(path)
+
+    changes = 0
+
+    try:
+        with open(os.open(path, os.O_RDONLY, dir_fd=dirfd), 'rb+') as f:
+            current = f.read()
+            if current != data:
+                changes |= FileChanges.CONTENTS
+
+            # The following cannot be skipped if we're changing file contents
+            # because we want accurate list of what has changed in file.
+
+            st = os.fstat(f.fileno())
+            if stat.S_IMODE(st.st_mode) != perms:
+                changes |= FileChanges.PERMS
+
+            if st.st_uid != uid:
+                changes |= FileChanges.UID
+
+            if st.st_gid != gid:
+                changes |= FileChanges.GID
+
+            if changes & (FileChanges.UID | FileChanges.GID):
+                os.fchown(f.fileno(), uid, gid)
+
+            if changes & FileChanges.PERMS:
+                os.fchmod(f.fileno(), perms)
+
+    except FileNotFoundError:
+        # Do not specify we're changing permissions on file
+        # because we're creating it new
+        changes = FileChanges.CONTENTS
+
+    if changes & FileChanges.CONTENTS:
+        with NamedTemporaryFile(mode='wb+', dir=parent_dir, delete=False) as tmp:
+            tmp.file.write(data)
+            tmp.file.flush()
+            os.fsync(tmp.file.fileno())
+            os.fchmod(tmp.file.fileno(), perms)
+            os.fchown(tmp.file.fileno(), uid, gid)
+            source_path = tmp.name
+
+        if dirfd is not None:
+            os.rename(source_path, path, src_dir_fd=dirfd, dst_dir_fd=dirfd)
 
-    changed = False
+        else:
+            os.rename(source_path, path)
```

Comment:
With regards to write, flush, fsync, fchmod, fchown,  how are exceptions or failures handled or passed to the caller?

Reply:
Passed back up to caller (keeping existing behavior).